SO-4645: replace Date with LocalDate in EffectiveTimes Java API

cmark opened the pull request:
Replace Date with LocalDate in all domain representation types.
Replace Date with LocalDate in all Java/REST API calls.

cmark committed:
SO-4645: replace Date with LocalDate in EffectiveTimes Java API

Replace Date with LocalDate in all domain representation types.
Replace Date with LocalDate in all Java/REST API calls.

apeteri commented on snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/components/SnomedPartialLoadingApiTest.java:
```diff
@@ -55,7 +53,7 @@ public void partialLoadingConceptApi() throws Exception {
 			// simple assertions to parse the ID as SCT ID and effective time as short date
 			assertNotNull(SnomedIdentifiers.create(hit[0]));
-			assertNotNull(EffectiveTimes.parse(hit[1], DateFormats.SHORT));
+			assertNotNull(Long.valueOf(hit[1]));
```
`Long.valueOf` never returns with `null`.

apeteri commented on tests/com.b2international.snowowl.test.commons/src/com/b2international/snowowl/test/commons/codesystem/CodeSystemVersionRestRequests.java:
```diff
@@ -108,34 +107,24 @@ public static ValidatableResponse getVersions(String shortName) {
 		if (!effectiveDates.isEmpty()) {
-			Date latestEffectiveDate = Dates.parse(effectiveDates.last(), DateFormats.SHORT);
+			LocalDate latestEffectiveDate = EffectiveTimes.parse(effectiveDates.last(), DateFormats.SHORT);
 			// XXX make sure we always use today or later dates, so all versions created in chronological order, even if some of the pre-imported content we are relying on are still in the past
 			// and adding one day to that historical version would mean a historical effective time version, which is unfortunate and can lead to inconsistencies in tests
```
The removed parts snapped `latestEffectiveDate` to today's date if it is in the past; the updated code starts with today's date and snaps to the latest effective date if it is in the future 😄 The comment should be updated and moved to the top of the method, since the process now starts there.

cmark replied:
Hmm, I'll check again, thanks. Also, I'll remove the same functionality from other test cases in other projects.

cmark committed:
SO-4645: fix SnomedPartialLoadingApiTest assertions

cmark committed:
SO-4645: properly calculate the next available effective time...

...in CodeSystemVersionRestRequests

Also, migrate to CodeSystemVersions model representation.

cmark committed:
SO-4645: migrate version API to CodeSystemVersion model representation

cmark committed:
SO-4645: add isLatest to CodeSystemVersion

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/version/CodeSystemVersionCreateRequest.java:
```diff
@@ -184,26 +184,25 @@ private void validateEffectiveTime(ServiceProvider context, CodeSystem codeSyste
-			.map(CodeSystemVersionEntry::getEffectiveDate)
-			.map(Instant::ofEpochMilli)
-			.orElse(Instant.EPOCH);
+			.findFirst()
```
This search request can now use `.one()` and `.first()` instead of `.all()` and `.stream().findFirst()`.

cmark committed:
SO-4645: fix version effectiveDate String to Date conversion in FHIR API

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemVersion.java:
```diff
@@ -95,36 +118,29 @@ public void setDescription(final String description) {
+	 */
+	@JsonIgnore
+	public boolean isLatestVersion() {
+		return MAIN_BRANCH.equals(getVersion());
```
Where do we build an artificial "latest" version like this? I couldn't find references in Eclipse for it.

cmark replied:
In other projects. This method is currently referenced in this repository.

cmark committed:
SO-4645: update launch configs

cmark committed:
SO-4645: fetch only the first version...

...when checking new version's effectiveTime validity

cmark committed:
SO-4645: make CodeSystemVersion serializable and final

apeteri commented on fhir/com.b2international.snowowl.fhir.core/src/com/b2international/snowowl/fhir/core/provider/FhirApiProvider.java:
```diff
@@ -91,26 +92,23 @@ protected final String getBranchPath(String version) {
-		return codeSystemOptional.orElseThrow(() -> 
-			new BadRequestException(String.format("Could not find corresponding version [%s] for logical id [%s].", logicalId.getBranchPath(), logicalId), "ValueSet.id"));
+			.first()
+			.orElseThrow(() -> new BadRequestException(String.format("Could not find corresponding version [%s] for logical id [%s].", logicalId.getBranchPath(), logicalId), "ValueSet.id"));
```
The exception message will use the output of `String.format` as a format string; I don't see where the last parameter fits into this picture.

```suggestion
			.orElseThrow(() -> new BadRequestException("Could not find corresponding version [%s] for logical id [%s].", logicalId.getBranchPath(), logicalId));
```

cmark committed:
Update fhir/com.b2international.snowowl.fhir.core/src/com/b2international/snowowl/fhir/core/provider/FhirApiProvider.java

Co-authored-by: András Péteri <apeteri@b2international.com>
